src/System.Collections.Generic.MultiValueDictionary/src/System/Collections/Generic/MultiValueDictionary.cs

```diff
@@ -0,0 +1,1102 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace System.Collections.Generic
+{
+    /// <summary>
+    /// A MultiValueDictionary can be viewed as a <see cref="IDictionary" /> that allows multiple 
+    /// values for any given unique key. While the MultiValueDictionary API is 
+    /// mostly the same as that of a regular <see cref="IDictionary" />, there is a distinction
+    /// in that getting the value for a key returns a <see cref="IReadOnlyCollection{TValue}" /> of values
+    /// rather than a single value associated with that key. Additionally, 
+    /// there is functionality to allow adding or removing more than a single
+    /// value at once. 
+    /// 
+    /// The MultiValueDictionary can also be viewed as a IReadOnlyDictionary&lt;TKey,IReadOnlyCollection&lt;TValue&gt;t&gt;
+    /// where the <see cref="IReadOnlyCollection{TValue}" /> is abstracted from the view of the programmer.
+    /// 
+    /// For a read-only MultiValueDictionary, see <see cref="ILookup{TKey, TValue}" />.
+    /// </summary>
+    /// <typeparam name="TKey">The type of the key.</typeparam>
+    /// <typeparam name="TValue">The type of the value.</typeparam>
+    public class MultiValueDictionary<TKey, TValue> :
+        IReadOnlyDictionary<TKey, IReadOnlyCollection<TValue>>
+    {
+        #region Variables
+        /*======================================================================
+        ** Variables
+        ======================================================================*/
+
+        /// <summary>
+        /// The private dictionary that this class effectively wraps around
+        /// </summary>
+        private Dictionary<TKey, InnerCollectionView> dictionary;
+
+        /// <summary>
+        /// The function to construct a new <see cref="ICollection{TValue}"/>
+        /// </summary>
+        /// <returns></returns>
+        private Func<ICollection<TValue>> NewCollectionFactory = () => new List<TValue>();
```

Comment:
The purpose of the NewCollectionFactory is to allow an instance of a MultiValueDictionary to be created that has a custom specified inner collection. I set it here to new List<TValue>() as the default, but I don't think it should be static. It would make sense that you may want a MultiValueDictionary of Lists as well as a MultiValueDictionary of HashSets in the same program. The static Factories starting on line 357 are an example of when this would be used.

 As an aside, if I remember correctly we also discussed removing the public NewCollectionFactory constructors in favor of a bool "AllowRepeatedValues" that would act as a simple function to choose between HashSet<TValue> and List<TValue> as the internal structure. It was decided on giving the user more control through the current pattern though.